### PR TITLE
hezt_model_py

### PR DESCRIPTION
**Pull request by Blessingwish**
import sys
import torch
import numpy as np
import mss
from pynput.keyboard import Controller
import pyautogui
from PyQt5.QtWidgets import QApplication, QWidget
from PyQt5.QtGui import QPainter, QPen, QColor
from PyQt5.QtCore import Qt, QTimer, QRect, QThread, pyqtSignal
import cv2
import time

# 定义目标类别
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')  # 使用 GPU 如果可用
classes = ['boss', 'ob', 'people', 'xianjin', 'jin']

# 加载训练好的 YOLOv5 模型
model = torch.hub.load('ultralytics/yolov5', 'custom', path='best.pt').to(device).eval()  # 设置为评估模式

# 屏幕捕获区域设置
monitor = {'top': 100, 'left': 0, 'width': 1920, 'height': 1080}  # 根据游戏窗口调整

# 初始化 pynput keyboard controller
keyboard_controller = Controller()

class DetectionThread(QThread):
    detection_signal = pyqtSignal(object)  # 信号，用于传递检测结果

    def run(self):
        with mss.mss() as sct:
            while True:
                # 捕获屏幕
                screenshot = sct.grab(monitor)
                img = np.array(screenshot)
                img = cv2.cvtColor(img, cv2.COLOR_RGBA2RGB)

                # 获取原始分辨率
                original_h, original_w, _ = img.shape

                # 调整图像大小为 YOLOv5 模型输入尺寸
                resized_img = cv2.resize(img, (480, 480))  # 调整为 480x480 以提升速度
                resized_h, resized_w = resized_img.shape[:2]

                # 使用 YOLOv5 模型进行检测
                results = model(resized_img)
                detections = results.xyxy[0].cpu().numpy()  # 获取检测框数据

                # 将检测框从 YOLOv5 输入尺寸映射回原始屏幕尺寸
                for det in detections:
                    det[0] = int(det[0] * original_w / resized_w)  # x1
                    det[1] = int(det[1] * original_h / resized_h)  # y1
                    det[2] = int(det[2] * original_w / resized_w)  # x2
                    det[3] = int(det[3] * original_h / resized_h)  # y2
                
                # 发射检测结果信号
                self.detection_signal.emit(detections)
                time.sleep(0.02)  # 减少延迟以提升识别速度

class Overlay(QWidget):
    def __init__(self):
        super().__init__()
        # 设置窗口为无边框、置顶、透明
        self.setWindowFlags(Qt.WindowStaysOnTopHint | Qt.FramelessWindowHint | Qt.Tool)
        self.setAttribute(Qt.WA_TransparentForMouseEvents, True)
        self.setAttribute(Qt.WA_NoSystemBackground, True)
        self.setAttribute(Qt.WA_TranslucentBackground, True)

        # 设置窗口大小为捕获区域大小
        self.setGeometry(monitor['left'], monitor['top'], monitor['width'], monitor['height'])

        # 初始化检测线程
        self.detection_thread = DetectionThread()
        self.detection_thread.detection_signal.connect(self.update_detections)
        self.detection_thread.start()

        # 定时器定期刷新界面
        self.timer = QTimer(self)
        self.timer.timeout.connect(self.repaint)
        self.timer.start(16)  # 60帧每秒 (1000/60 ≈ 16ms)

        # 用于保存检测结果
        self.detections = []
        self.attacking = False  # 是否正在攻击
        self.switched_weapon = False  # 是否已切换武器

    def update_detections(self, detections):
        """更新检测结果"""
        print("Updated detections:", detections)  # 打印调试信息
        self.detections = detections
        self.handle_boss_detection()

    def handle_boss_detection(self):
        """检测到 boss 时模拟鼠标按住左键自动攻击"""
        for det in self.detections:
            confidence = det[4]  # 置信度
            if confidence > 0.3:  # 设置置信度阈值
                label = int(det[5])  # 获取类别标签
                class_name = classes[label]

                if class_name == 'boss':
                    # 获取 boss 的中心坐标
                    x1, y1, x2, y2 = map(int, det[:4])
                    boss_center_x = (x1 + x2) // 2
                    boss_center_y = (y1 + y2) // 2

                    # 计算屏幕坐标
                    screen_x = monitor['left'] + boss_center_x
                    screen_y = monitor['top'] + boss_center_y

                    # 移动鼠标到 boss 位置
                    pyautogui.moveTo(screen_x, screen_y)

                    # 切换武器（仅切换一次）
                    if not self.switched_weapon:
                        keyboard_controller.press('1')  # 按下数字键 1
                        time.sleep(0.1)
                        keyboard_controller.release('1')  # 释放数字键 1
                        self.switched_weapon = True

                    # 如果未开始攻击，按住左键
                    if not self.attacking:
                        pyautogui.mouseDown()  # 按住鼠标左键
                        self.attacking = True
                    return

        # 如果没有检测到 boss，则释放鼠标左键并重置武器状态
        if self.attacking:
            pyautogui.mouseUp()  # 释放鼠标左键
            self.attacking = False
        self.switched_weapon = False  # 重置武器切换状态

    def paintEvent(self, event):
        """绘制检测框"""
        if hasattr(self, 'detections'):
            painter = QPainter(self)
            pen = QPen(QColor(255, 0, 0), 2)  # 红色边框，线宽 2
            painter.setPen(pen)

            for det in self.detections:
                confidence = det[4]
                if confidence > 0.3:  # 置信度阈值
                    x1, y1, x2, y2 = map(int, det[:4])
                    label = int(det[5])
                    class_name = classes[label]

                    # 绘制矩形框和类别标签
                    painter.drawRect(QRect(x1, y1, x2 - x1, y2 - y1))
                    painter.drawText(x1, y1 - 10, f'{class_name} {confidence:.2f}')
            painter.end()

if __name__ == '__main__':
    app = QApplication(sys.argv)
    overlay = Overlay()
    overlay.show()
    sys.exit(app.exec_())

### CODE REVIEW

**Comment by donghaiwang** [comment]
不是添加文件，是在原来的文件基础上添加注释。
